sync/src/relayer/transaction_hashes_process.rs

```diff
@@ -0,0 +1,111 @@
+use crate::relayer::compact_block::TransactionHashes;
+use crate::relayer::Relayer;
+use ckb_core::transaction::ProposalShortId;
+use ckb_logger::{debug_target, trace_target};
+use ckb_network::PeerIndex;
+use ckb_protocol::RelayTransactionHashes as FbsRelayTransactionHashes;
+use ckb_store::ChainStore;
+use failure::Error as FailureError;
+use numext_fixed_hash::H256;
+use std::convert::TryInto;
+
+pub struct TransactionHashesProcess<'a, CS> {
+    message: &'a FbsRelayTransactionHashes<'a>,
+    relayer: &'a Relayer<CS>,
+    peer: PeerIndex,
+}
+
+impl<'a, CS: ChainStore + 'static> TransactionHashesProcess<'a, CS> {
+    pub fn new(
+        message: &'a FbsRelayTransactionHashes,
+        relayer: &'a Relayer<CS>,
+        peer: PeerIndex,
+    ) -> Self {
+        TransactionHashesProcess {
+            message,
+            relayer,
+            peer,
+        }
+    }
+
+    pub fn execute(self) -> Result<(), FailureError> {
+        let transaction_hashes: TransactionHashes = (*self.message).try_into()?;
+
+        let mut transit_hashes: Vec<H256> = {
+            let tx_filter = self.relayer.shared().tx_filter();
+            transaction_hashes
+                .hashes
+                .into_iter()
+                .filter(|tx_hash| {
+                    if tx_filter.contains(&tx_hash) {
+                        debug_target!(
+                            crate::LOG_TARGET_RELAY,
+                            "transaction({:#x}) from {} already known, ignore it",
+                            tx_hash,
+                            self.peer,
+                        );
+                        false
+                    } else {
+                        true
+                    }
+                })
+                .collect()
+        };
+
+        let mut knowned = Vec::with_capacity(transit_hashes.len());
+        {
+            let state = self.relayer.shared.lock_chain_state();
+            let tx_pool = state.tx_pool();
+
+            transit_hashes.retain(|tx_hash| {
+                let short_id = ProposalShortId::from_tx_hash(&tx_hash);
+                if tx_pool.contains_tx(&short_id) {
+                    trace_target!(
+                        crate::LOG_TARGET_RELAY,
+                        "transaction({:#x}) from {} already in transaction pool, ignore it",
+                        tx_hash,
+                        self.peer,
+                    );
+                    knowned.push(tx_hash.to_owned());
+                    false
+                } else {
+                    true
+                }
+            })
+        }
+
+        {
+            self.relayer.shared().mark_as_known_txs(knowned);
```

Comment:
I think it is unnecessary to mark known txs here. It should very rarely that "`tx_pool` has a tx but `tx_filter` not". So I would suggest removing this logic or add a log when `!knowned.is_empty()`

Reply:
I am not sure, just keep the logic as before.

Reply:
agree, the filter is a false negative.